Fix popover not calling onClose on unmount

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes  https://github.com/WordPress/gutenberg/issues/71281, which is a blocking issue in https://github.com/WordPress/gutenberg/pull/71163#issuecomment-3191252617

<!-- In a few words, what is the PR actually doing? -->

When a popover unmounts, it should call its related onClose, if provided.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
It makes sense - the popover is closing, so the onClose should be called.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Remove cancelling the onBlur check when unmounting the popover.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

- All popovers should retain the same behavior as they currently do
- To test what this fixes, checkout this branch: `git checkout test/onclose-popover-unmount`. This is a combination of https://github.com/WordPress/gutenberg/pull/71163/ but with [this commit removed that tried to fix this issue with a prop](https://github.com/WordPress/gutenberg/pull/71163/commits/dbeb8b8156153dd010b842959d10c30d6b1d1071).
- Click a navigation block in the site editor to show the appender
- Open the sidebar 
- Click the appender for the navigation block within the sidebar
- The popover should open and there should be a "Select Page" placeholder in the canvas for where the new link will go
- Click into the canvas
- The popover should be removed and the Select Page should be removed (through the onClose that is called when the popover unmounts).

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

**Before**

https://github.com/user-attachments/assets/44f1933f-d568-4689-8c2f-721c7cea9407

**After**

https://github.com/user-attachments/assets/8d1c62ea-22ae-4c23-bc06-5d64e1c033a9

jeryj committed:
Run onClose when unmounting component

There was a cancelBlurCheck running on unmount, which prevented the onClose for the popover from running when the popover was unmounted.

aduth reviewed (commented):
Based on the discussion in #71163, my hope is we could get rid of the timeout altogether, and instead check `event.relatedTarget` synchronously as part of the `onBlur` handler.

Roughly something like:

```tsx
const onBlur = useCallback( ( event ) => {
  if ( ! event.target.contains( event.relatedTarget ) ) {
    onFocusOutside();
  }
} );

return { onBlur };
```

How feasible would that be?

jeryj commented:
@ciampo - It fixes the bug we're experiencing in https://github.com/WordPress/gutenberg/pull/71163#discussion_r2278741629 where the popover onClose isn't being called, even though the popover is being closed.

@aduth - I'm looking at the implmentation using just the `onBlur`: `! event.target.contains( event.relatedTarget )`. I think I'll have to pass a ref into it though. I'm seeing `event.target` being logged as the element that is losing focus, and `event.relatedTarget` as the element receiving focus. The container seems hard to parse out since it's all using portals to populate the popover. Is passing a ref from useDialog a viable option here, or would that create too many back compat issues?

aduth commented:
@jeryj Thanks for digging into that, and apologies on behalf of my past self for not leaving more context on the blur cancellation behavior 😅 My sense was that it was a browser-specific quirk ("A blur event's relatedTarget is not reliable in IE11" in #3235), hence hammering so much on the potential removal now that we no longer support IE11.

>Long story short: Since the unmount cancelling the timeout is causing a bug for us, and I don't see a clear reason why it was added, I feel comfortable with removing the unmount hook if we don't see it causing any new problems.

Considering everything you mentioned, I agree that it's best to be consistent in how we're triggering this callback asynchronously, like you have in the proposed implementation.

Have you been able to do some testing on potential impact to other components using this? I see it's not too many, with most of its usage going through the higher-order component (`ComboBox` uses `withFocusOutside`) or higher-level abstractions (`useDialog`, in turn used primarily in `Popover`, but also in `EntitiesSavedStates`).

jeryj commented:
> What do you think? Is this taking things too far?

I think it may be riskier, as the normal flow and unmount flow would be separate. For example:

- `queueMicrotask` might not run at the same time in the cycle as `setTimeout( () => {}, 0 );`
- The original event and synthetic event might not be identical

So, I think the risk of creating a new system to handle unmount that is separate from the default flow is greater than having a  potential uncleared timeout. That's my untested opinion. Very open to being wrong here :)

aduth reviewed (commented):
Personally, I'm generally in favor of this approach if it helps unblock #71163, even if there's additional opportunities to explore with synchronous callbacks, which could be done as iterative improvements. I also see some value in being consistent across the board in how the hook calls the callback (rather than a mix of some synchronous and some asynchronous callbacks).

While I think this looks good, a couple questions that come to mind in last-minute thoughts about the approach:

- Are there potential issues that could occur if the popover is unmounted because the parent component is also unmounted? I'm imagining `onClose` being called on some parent component that's been unmounted and that parent component callback taking some action like setting state, and causing warnings like "Can't perform a React state update on an unmounted component."
- If this is fixing something and we're relying on tests to give confidence in this not causing regressions, should there not also be some test code revisions here to validate the expected change in behavior?

jeryj commented:
> Are there potential issues that could occur if the popover is unmounted because the parent component is also unmounted? I'm imagining onClose being called on some parent component that's been unmounted and that parent component callback taking some action like setting state, and causing warnings like "Can't perform a React state update on an unmounted component."

If the parent component using the useFocusCallback is unmounted, would [this effect that cancels the blur check](https://github.com/WordPress/gutenberg/pull/71252/files#diff-ff86e23e91ebf4904d9c61103345eb971caecf569334227d6943e359c67b16e2R90-R95) be triggered, which would cancel the callback? Something like:
- Parent is unmounted
- `onFocusOutside` (the callback) becomes null? (I'm not sure this is what happens)
- The `useEffect` with the `onFocusOutside` dependency changes, cancelling the blur check.
- `setTimeout` doesn't run. 

> If this is fixing something and we're relying on tests to give confidence in this not causing regressions, should there not also be some test code revisions here to validate the expected change in behavior?

Yup! That should be the case. I'll get a test that demonstrate the callback runs on unmount.

getdave reviewed (approved):
I'm going to take a leap and approve this one. 

**Update**: I added some unit tests coverage, although I'd appreciate a confidence check that it's covering what I think it's covering 😅 

I have removed all custom implementation from https://github.com/WordPress/gutenberg/pull/71163. I then applied this PR to that one. I then added a temporary `setState` call to the `onClose` callback to see if the unmount of the parent would cause React to fire its warning:

```jsx
<LinkUI
			clientId={ insertedBlock?.clientId }
			link={ insertedBlock?.attributes }
			onBlockInsert={ handleSetInsertedBlock }
			onClose={ () => {
				// Use cleanup function
				cleanupInsertedBlock();
				setRandomState( 'crash all the things!' );
			} }
....
```

I did not see any warning. I don't think this is necessarily conclusive proof that we'll never encounter such a scenario but it's a good indicator.

There also seems to be a general consensus that the chances of bugs are relatively low and that we should ship and iterate on improve the approach if issues are uncovered.

If anyone disagrees please let us know, otherwise I'll let @jeryj press the merge button when he's online later today.

getdave committed:
Add test coverage for hook

aduth commented on packages/compose/src/hooks/use-focus-outside/test/index.js:
```diff
@@ -130,11 +130,19 @@ describe( 'useFocusOutside', () => {
 		unmount();
 
-		expect( mockOnFocusOutside ).not.toHaveBeenCalled();
+		// Wait for the queued timeout to execute
+		await new Promise( ( resolve ) => setTimeout( resolve, 10 ) );
+
+		expect( mockOnFocusOutside ).toHaveBeenCalled();
```
I'd discourage a real-time sleep like this, both because it's pretty fragile, and because it artificially prolongs the test runtime (although not by much in this case).

Could we do something where we set up the mock implementation to finish the test? Either using [the `done` callback](https://jestjs.io/docs/asynchronous#callbacks), or maybe something with [`Promise.withResolvers`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers), e.g.

```ts
const { promise, resolve } = Promise.withResolvers();
mockOnFocusOutside.mockImplementation( resolve );
unmount();
await promise;
```

Mamaduka replied:
The `withResolvers` is such a nice addition to the promises ⭐

getdave replied:
TIL about `withResolvers()`. Thanks!

getdave replied:
Looks like `Promise.withResolvers()` is only available in Node.js 22+, but the project requires Node.js >=20.10.0. This was leading to unit test failures on CI.

I've reverted to using the traditional manual deferred promise pattern for better compatibility.

aduth reviewed (approved):
> > Can't perform a React state update on an unmounted component.
>
>React no longer triggers setState warnings, since version 18.

Thanks @Mamaduka , I overlooked that. Based on [the context of that change](https://github.com/reactwg/react-18/discussions/82) and some more thought about my previous concerns, the sense I have is that: If a component using `useFocusOutside` unmounts because its parent unmounts, any callbacks on that parent will still be called, even though its unmounted. That could cause some unintentional side-effects depending on what those callbacks are doing, but in the case of React-related effects like `setState`, they'll be effectively treated as a no-op (and no longer log as of v18). It feels like a pretty low risk overall.

getdave committed:
Improve useFocusOutside test with Promise.withResolvers()

Replace fragile setTimeout approach with Promise.withResolvers() to make
the test more reliable and faster. The test now waits for the actual
callback to be called rather than an arbitrary timeout.

Addresses feedback from PR #71252 review.

getdave committed:
Fix Node.js compatibility: replace Promise.withResolvers() with manual deferred pattern

Promise.withResolvers() is only available in Node.js 22+, but the project
requires Node.js >=20.10.0. Use the traditional manual deferred promise
pattern for better compatibility.

This maintains the same improved test reliability while ensuring
compatibility with the project's Node.js version requirements.
